### PR TITLE
Feature/128 Package updates back to workable.

### PR DESCRIPTION
**Pull request by owenmorrill**

### CODE REVIEW

**Comment by owenmorrill** [comment]
I had to revert back to using .getJSON instead of fetch because the response from fetch wasn't parsing out into what I needed. There's probably a better way to do it, but I reverted and got it working. Next step would be to figure out what fetch is doing and make that work.

**Comment by alexfinnarn** [comment]
On https://github.com/CuBoulder/lil_shrugger/blob/763764b8bd88f963ed51b09c206591b26c1cdcca/update-sites.html#L41

I think that making that a value instead of a placeholder would be more valuable, especially if that query is used a lot. I imagine this search could use queries from the new/forth-coming Atlas "/query" endpoint, but for now, it'd be nice if I could just take code asset IDs from the "/code" page and then paste them into that initial query. 

screen shot 2017-08-10 at 10 13 47 am

In the screenshot above, if I select one "cu_newsletter_bundle" then any others in the list will also be selected. They appear to be at the same hash, and when I look at the code endpoint, there is only one Newsletter Bundle code asset. I think what is happening is that any site that has that bundle is getting it listed without de-duplicating them in the list. 

screen shot 2017-08-10 at 10 22 31 am

In the screenshot, the added/removed packages and sites list is kind of hard to read. I think they used to be listed in bullet points and that format would be more useful for looking at these query results. I just queried for a Drupal core asset to see how it looked with a lot of sites.

**Comment by alexfinnarn** [comment]
As far as the Atlas requests go vs. $.getJSON(), the atlasRequest() function returns a paged response to get more than 500 results, which is the default max. 

```
// $.getJSON returns the full response with _items in it but doesn't fetch second page.
that.sites = data

// atlasRequest returns an array of data that needs to be looped through for display. 
that.sites = data[0] + data[1] + ... // Some concatenating.
```

File: src/js/update-sites.js [hunk 1]
```diff
@@ -20,26 +22,29 @@ let vm = new Vue({
...
   methods: {
     getSites(siteQuery) {
       let that = this;
-      let endpoint = '/sites';
+      let endpoint = 'sites';
+      console.log(packageBaseURL+endpoint+siteQuery);
 
-      atlasRequest(baseURL, endpoint, query = siteQuery, method = 'GET', body = null, etag = null).then(data => that.sites = data);
+      atlasRequest(packageBaseURL, endpoint, query = siteQuery, method = 'GET', body = null, etag = null).then(data => that.sites = data);
+      setTimeout(remDup, 50);
```
Comment (alexfinnarn) [inline src/js/update-sites.js]: I would filter the list of sites on line 33 in the thenable before assigning to `that.sites`.

File: src/js/update-sites.js [hunk 2]
```diff
@@ -3,12 +3,14 @@
...
  */
-let link = document.querySelector('link[href="src/partials/listing.html"]');
-let content = link.import;
-let el = content.querySelector('script');
-document.querySelector('body').appendChild(el.cloneNode(true));
+let packageBaseURL = store.state.atlasEnvironments[store.state.env];
 
-let baseURL = store.state.atlasEnvironments[store.state.env];
+function remDup() {
+  $("#packages-to-add option").val(function(idx, val) {
```
Comment (alexfinnarn) [inline src/js/update-sites.js]: There should be no DOM manipulation in this application done by us. Vue nicely takes care of that, and so filtering the data before rendering it is the best option.

File: src/js/update-sites.js [hunk 3]
```diff
@@ -20,26 +22,29 @@ let vm = new Vue({
...
-      let endpoint = '/code';
+      let endpoint = 'code';
+      console.log(packageBaseURL+endpoint);
 
-      atlasRequest(baseURL, endpoint, query = '', method = 'GET', body = null, etag = null).then(data => that.codes = data);
+      atlasRequest(packageBaseURL, endpoint, query = '', method = 'GET', body = null, etag = null).then(data => that.codes = data);
     },
     sendToAtlas() {
-      let siteOriginalPackages = this.sites._items;
+      let siteOriginalPackages = this.sites[0];
```
Comment (alexfinnarn) [inline src/js/update-sites.js]: What if there is `this.sites[1]`? If there are more than 500 sites returned in the query then another element in the array will exist. You can test this by looking for the current core or profile code asset.

File: src/js/update-sites.js [hunk 4]
```diff
@@ -86,7 +91,7 @@ let vm = new Vue({
 
     },
     codeName(index) {
-      let codeItem = vm.codes._items;
+      let codeItem = vm.codes[0];
```
Comment (alexfinnarn) [inline src/js/update-sites.js]: Same as above...there might be more than one element in the array.

File: src/js/update-sites.js (hunk 3 above)
Comment (owenmorrill) [inline src/js/update-sites.js]: This is what I'm getting back with {{sites}}:

`[ [ { "code": { "profile": "59910dac3644752d498d79da", "core": "59836e68364475026576d2c3", "package": [ "588b90a93644752164c7beae", "588b90ab36447521657fd40f", "588b90ab36447521c5af51e4", "588b90ad364475217a9fcf32", "588b90ad36447521657fd411", "59398ded3644755921511d13" ] }, "created_by": "jafu6031", "statistics": "588f8d113644753e810ef343", "modified_by": "KERE7580", "_deleted": false, "_links": { "self": { "href": "sites/588f8d103644753e838b273b", "title": "site" } }, "sid": "p1c4038dad8c", "type": "express", "status": "launched", "_latest_version": 39, "f5only": false, "db_key": "67414141414142596a343052504a5944424176716d4171583269564454446f70337252486c4a50396231796963736d38344478674a47627a315864626f334c506652386c456d4c536f323469347a76304c773956656457516f4343344556495746706b69366d6e3638597474317a3649455667356a585f62794e702d2d36427777314d62373549412d346271", "path": "webcentral", "pool": "poolb-express", "_updated": "2017-08-14 03:09:57 GMT", "update_group": 11, "settings": { "page_cache_maximum_age": 3600 }, "_version": 39, "import_from_inventory": true, "_created": "2017-01-30 18:59:28 GMT", "_id": "588f8d103644753e838b273b", "_etag": "505e056b83d498914cd5366d8c63ff291996efc6" }, { "code": { "profile": "59910dac3644752d498d79da", "core": "59836e68364475026576d2c3", "package": [ "588b90a036447521fd349bc2", "588b90a036447521c5af51d8", "588b90a13644752166c62225", "588b90a136447521fd349bc4", "588b90a136447521633019af", "588b90a236447521c5af51da", "588b90a23644752166c62227", "588b90a236447521fd349bc6", "588b90ab36447521c5af51e4", "589387203644754f7d3295d3", "58a615fc364475514798f9b7", "58b995c53644755c0c936f00", "58e7f86b3644750838d3a508", "592da2d93644756b06fae298", "592eec973644750954a2f28a", "592eedb136447569d6a52639", "5931c1d236447550e4d39a3e", "59398ded3644755921511d13", "5963d4fb36447508b0b83880", "596e1a65364475781b834847", "5989cf873644755ff4194732", "588b90ab36447521c5af51e4", "589387203644754f7d3295d3", "588b90ab36447521c5af51e4", "589387203644754f7d3295d3", "588b90ab36447521c5af51e4", "58a713c0364475207b295d51", "588b90ab36447521c5af51e4", "589387203644754f7d3295d3", "588b90ab36447521c5af51e4", "58a713c0364475207b295d51", "588b90a036447521fd349bc2", "588b90a736447521fd349bcc" ] }, "created_by": "jafu6031", "statistics": "588fae5436447566b3b45f04", "modified_by": "JEOR0980", "_deleted": false, "_links": { "self": { "href": "sites/588fae543644756cfc235068", "title": "site" } }, "sid": "p15ac50597de", "type": "express", "status": "launched", "_latest_version": 28, "f5only": false, "db_key": "67414141414142596a36355565564666434e5a415479375574563336536b486774794633617169706b344847643169744437487730735f694c596f49443834716d754345783841686b56704c39384864734d564971464a6468626133706d434d63722d446c705132776e434257773867423541644537415578664d46647278356238564b726d354158305964", "path": "cas", "pool": "poolb-express", "_updated": "2017-08-14 21:41:05 GMT", "update_group": 11, "settings": { "page_cache_maximum_age": 3600 }, "_version": 28, "import_from_inventory": true, "_created": "2017-01-30 21:21:24 GMT", "_id": "588fae543644756cfc235068", "_etag": "0b8e4f75b50d0a4c8db608cbdb2890957e790298" }, { "code": { "profile": "59910dac3644752d498d79da", "core": "59836e68364475026576d2c3", "package": [ "588b90ab36447521c5af51e4", "58a615fc364475514798f9b7", "58a713c0364475207b295d51", "588b90ab36447521c5af51e4", "58a713c0364475207b295d51" ] }, "created_by": "osr_atlas_client", "statistics": "589538ac3644757dd2a371fc", "modified_by": "KERE7580", "_deleted": false, "_links": { "self": { "href": "sites/589538ac3644754f0a53334c", "title": "site" } }, "sid": "p158837d3f1e", "type": "express", "status": "launched", "_latest_version": 43, "f5only": false, "db_key": "67414141414142596c5469734d6c3155324c4a645f437359515833375146635a7873513161534f4e75393969416266616d62516c6978784c38556c7179485552586d4357344e5253393770534443304855685567744a535832496a2d75386671585a37736b66446f4354756950634f386268384b3064445639574b542d472d72783278526f2d303052333865", "path": "crowdfunding", "pool": "poolb-express", "_updated": "2017-08-14 03:10:09 GMT", "dates": { "launched": "2017-02-13 23:07:09 GMT", "created": "2017-02-04 02:12:59 GMT" }, "update_group": 0, "settings": { "page_cache_maximum_age": 3600 }, "_version": 43, "import_from_inventory": false, "_created": "2017-02-04 02:12:59 GMT", "_id": "589538ac3644754f0a53334c", "_etag": "8c6c6fef349bfd48eaea55b30b0f60b2ce7d6884" } ] ]`

So the [0] is actually bypassing some random array to get to the nested array of the actual sites, right?

File: src/js/update-sites.js [hunk 5]
```diff
@@ -20,35 +19,49 @@ let vm = new Vue({
...
+        function setPackages() {
+          let singlePackage = [];
+          for(site of that.sites) {
+            for(single of site.code.package) {
+              singlePackage.push(single);
+            }
+          }
+          var names = singlePackage;
+          var uniqueNames = [];
+          $.each(names, function(i, el){
```
Comment (alexfinnarn) [inline src/js/update-sites.js]: I think `$` is jQuery? If this is just a foreach loop then you should be able to do `names.forEach(function(i,el){...});`

File: src/js/update-sites.js [hunk 6]
```diff
@@ -20,35 +19,49 @@ let vm = new Vue({
...
+            if($.inArray(el, uniqueNames) === -1) uniqueNames.push(el);
```
Comment (alexfinnarn) [inline src/js/update-sites.js]: `$.inArray` could be `uniqueNames.findIndex()`.

**Comment by alexfinnarn** [comment]
Also, when I tried to add a package that was already in the site record, two of the same code assets were inserted. Deduping the packages array before sending should fix that.

**Comment by alexfinnarn** [comment]
closing since the code has changed quite a bit.

**Pull request closed without merging**